fql/setup.js

```diff
@@ -0,0 +1,148 @@
+  // Index contains the Canvas ref (that's the default behavior and could be omitted)
+  values: [
+    { field: ['ref'] },
+  ],
+});
+
+// ---------------------- Step 4: Create roles ----------------------
+
+// The "Editor" role is assigned to all authenticated users
+// It is automatically assigned when a user is authenticated, because it defines "membership" to the Users collection
+// It is secure because the token is generated upon login on the server-side and stored in a "httpOnly" cookie that can only be read/written on the server-side
+// The token is specific to the user and is used on the frontend
+// The token only allows the user to read/write documents that belongs to him
+CreateRole({
+  name: 'Editor',
+  // All users should be editors (will apply to authenticated users only).
+  membership: {
+    resource: Collection('Users'),
+  },
+  privileges: [
+    {
+      // Editors need read access to the canvas_by_owner index to find their own canvas
+      resource: Index('canvas_by_owner'),
+      actions: {
+        read: true,
+      },
+    },
+    {
+      resource: Collection('Canvas'),
+      actions: {
+        // Editors should be able to read (+ history) of Canvas documents that belongs to them.
+        read: Query(
+          Lambda('ref', Equals(
+            CurrentIdentity(),
+            Select(['data', 'owner'], Get(Var('ref'))),
+          )),
+        ),
+        history_read: Query(
+          Lambda('ref', Equals(
+            CurrentIdentity(),
+            Select(['data', 'owner'], Get(Var('ref'))),
+          )),
+        ),
+        // Editors should be able to edit only Canvas documents that belongs to them
+        write: Lambda(
+          ["oldData", "newData", "ref"],
+          And(
+            // The owner in the current data (before writing them) must be the current user
+            Equals(
+              CurrentIdentity(),
+              Select(["data", "owner"], Var("oldData"))
+            ),
+            // The owner must not change
+            Equals(
+              Select(["data", "owner"], Var("oldData")),
+              Select(["data", "owner"], Var("newData"))
+            )
+          )
+        ),
+        // Editors should be able to create only Canvas documents that belongs to them
+        create: Lambda("values", Equals(
+          CurrentIdentity(),
+          Select(["data", "owner"], Var("values")))
+        ),
+      },
+    },
+  ],
+});
+
+// The "Public" role is assigned to anyone who isn't authenticated
+// It doesn't use "membership" (unlike "Editor" role) but a token created manually that doesn't expire
+// It is secure because the token only grant access to the special document of id "1", which is shared amongst all guests
+// Guests can only read/write this particular document and not any other
+CreateRole({
+  name: 'Public',
+  // The public role is meant to be used to generate a token which allows anyone (unauthenticated users) to update the canvas
+  membership: {},
```

Comment:
If you create tokens, you must pass an instance ref, given this, that role will be discarded because it doesn't contains any membership predicate.

Membership-less roles are meant to be used with JWT tokens (see Access Providers)

Reply:
I guess I used the wrong wording, the Public role isn't using a token, but a key.

![image](https://user-images.githubusercontent.com/3807458/111748990-7933b780-8891-11eb-9047-ef2b89869e9f.png)